### PR TITLE
feat(sub-group deploy): add sub-group deployment feature

### PR DESCRIPTION
**Pull request by devnaruka**
**Issue #, if available:**

**Description of changes:**

1. Allow subgroup deployment to override parent group components.
2. Enforce deployments to be processed based on creation time sequence. All stale deployments would be rejected.
3. Fleet status update for rejection status would have detailed status.
4. Maintain last rejected deployment record similar to success and failed deployments.
5. Update unit tests for group to last deployment topic lookup in deployment config.
6. Add integration tests for subgroup deployments covering siblings, hierarchical, nested and out-of-order deployment scenarios.

**Why is this change necessary:**
To enable customers troubleshoot and help them reduce their time to recover by empowering them with tools like sub-group deployments. Sub-group deployments let's customers override part of their fleet with the changes they want to:
1. retry a failed deployment configuration
2. fix a bug which occurs only in limited device configurations

**How was this change tested:**
- [x] Updated or added new unit tests.
- [x] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [x] I confirm that the change is backwards compatible.
- [x] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java [hunk 1]
```diff
@@ -148,4 +150,15 @@ public static void deleteWindowsTestUser(String username) throws IOException, In
         // Delete test user's home directory
         FileUtils.deleteQuietly(new File("C:\\Users\\" + username));
     }
+
+
+    @SuppressWarnings({"PMD.SignatureDeclareThrowsException"})
+    public static void overrideInstanceField(Object target, String field, Object value) throws Exception {
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java:156]: what? Why does this exist

File: src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java [hunk 2]
```diff
@@ -111,6 +114,20 @@ public DeploymentResult call() throws InterruptedException {
                     .kv("Deployment service config", deploymentServiceConfig.toPOJO().toString())
                     .log("Starting deployment task");
 
+            // Enforce deployment sequence validation using timestamps:
+            // If we have already processed a more recent deployment for a thing group, then we will drop
+            // this deployment and only process it if it is newer (or same) deployment for that thing group.
+            if (isStaleDeployment(deploymentDocument)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java:120]: can we do this in a different PR. Stale deployments and subgroups don't need to be tied.

File: src/main/java/com/aws/greengrass/deployment/model/DeploymentDocument.java [hunk 3]
```diff
@@ -82,6 +89,17 @@ public class DeploymentDocument {
...
 
+
+    /**
+     * For sub-group deployments root group name is used otherwise group name.
+     *
+     * @return if available root group name, otherwise group name
+     */
+    @JsonGetter("GroupName")
+    public String getGroupName() {
+        return onBehalfOf == null || onBehalfOf.isEmpty() ? groupName : onBehalfOf;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/model/DeploymentDocument.java:100]: You can use `Utils.isEmpty()`

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java [hunk 4]
```diff
@@ -343,6 +345,29 @@ private void finishCurrentDeployment() throws InterruptedException {
...
+                    deploymentStatusKeeper.persistAndPublishDeploymentStatus(deploymentId, ggDeploymentId,
+                            configurationArn, type, JobStatus.REJECTED.toString(), statusDetails, rootPackages);
+
+                    if (currentDeploymentTaskMetadata.getDeploymentTask() instanceof KernelUpdateDeploymentTask) {
+                        try {
+                            kernel.getContext().get(KernelAlternatives.class).rollbackCompletes();
+                        } catch (IOException e) {
+                            logger.atError().log("Failed to reset Nucleus rollback directory", e);
+                        }
+                    }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentService.java:369]: why are we doing this? We never ran the deployment because we rejected it, so why are we dealing with rollbacks?

File: src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java [hunk 5]
```diff
@@ -41,6 +41,7 @@ public class DeploymentDirectoryManager {
 
     private static final String PREVIOUS_SUCCESS_LINK = "previous-success";
     private static final String PREVIOUS_FAILURE_LINK = "previous-failure";
+    private static final String PREVIOUS_REJECTED_LINK = "previous-rejected";
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java:44]: I don't think we really want a directory for rejected deployments

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java (hunk 4 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DeploymentService.java:369]: Makes sense. I will remove it!

File: src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java [hunk 6]
```diff
@@ -67,7 +67,7 @@ class DeploymentTaskTest {
     private static Context context;
     private final DeploymentDocument deploymentDocument =
             DeploymentDocument.builder().deploymentId("TestDeployment").timestamp(System.currentTimeMillis())
-                    .groupName(DeploymentDocumentConverter.LOCAL_DEPLOYMENT_GROUP_NAME).build();
+                    .groupName(DeploymentDocumentConverter.THING_GROUP_RESOURCE_NAME_PREFIX + DeploymentDocumentConverter.LOCAL_DEPLOYMENT_GROUP_NAME).build();
```
[the remark this answers is not in the document]
Comment (devnaruka) [reply src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java:70]: No, it doesn't break local deployments. This was just for consistency. I will revert it back!

File: src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java (hunk 2 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java:120]: That would be tough now!

File: src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java (hunk 1 above)
Comment (devnaruka) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java:156]: oh, I didn't find a good way to capture FSS status updates in my integ tests to check rejection status. I had to override the deployment config with the spy version. I need this to replace the private deploymentConfig value.

Maybe you can suggest a better alternative, as my knowledge about FSS service setup is limited.

File: src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java (hunk 5 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java:44]: I can remove this as well. I thought we are maintaining records of both success and failure, then we might also be interested in storing rejected too.

File: src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java (hunk 2 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java]: I have split the commit and removed the stale deployment rejection. I will publish a separate CR for that.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java (hunk 1 above)
Comment (devnaruka) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/BaseITCase.java]: This has been removed and we can circle back on this for listening to fss update messages for rejection status.

File: src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java [hunk 7]
```diff
@@ -85,10 +85,10 @@ public void persistLastSuccessfulDeployment() {
     private void persistPointerToLastFinishedDeployment(Path symlink) {
         logger.atInfo().kv(LINK_LOG_KEY, symlink).log("Persist link to last deployment");
         try {
-            Path deploymentPath = getDeploymentDirectoryPath();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java:88]: just undo all the changes in this file

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java [hunk 8]
```diff
@@ -363,13 +363,6 @@ private void finishCurrentDeployment() throws InterruptedException {
                     deploymentStatusKeeper.persistAndPublishDeploymentStatus(deploymentId, ggDeploymentId,
                             configurationArn, type, JobStatus.FAILED.toString(), statusDetails, rootPackages);
 
-                    if (currentDeploymentTaskMetadata.getDeploymentTask() instanceof KernelUpdateDeploymentTask) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentService.java:366]: hold on, why is this removed now? We didn't want if for REJECTED, but we should keep it here, I'm sure it is doing something useful

File: src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java [hunk 9]
```diff
@@ -92,6 +92,7 @@ class DeploymentTaskTest {
 
     private Topics mockGroupToRootConfig;
     private Topics mockGroupMembership;
+    private Topics mockLastDeploymentTopics;
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java:95]: unused?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/SubGroupDeploymentIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+    }
+
+    @AfterEach
+    void after() {
+        if (kernel != null) {
+            kernel.shutdown();
+        }
+    }
+
+    @Test
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/SubGroupDeploymentIntegrationTest.java:130]: All these tests are going to be very slow because they're all starting a new kernel and running deployments. I don't think that most of these are necessary for validating the functionality.

I'd suggest that you collapse some of these test scenarios into single test cases. 

The most critical test for me is
1. Deploy to parent group
2. Deploy to subgroup with a deployment which would conflict with the parent if the subgroup logic wasn't implemented
3. Deploy to sibling subgroup with a deployment which would conflict with the parent and the previous subgroup
4. Deploy to parent group again and verify it overrides what the subgroups brought

Comment (devnaruka) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/SubGroupDeploymentIntegrationTest.java:130]: Ok, no problem. I will remove all other tests then!

File: src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java (hunk 9 above)
Comment (devnaruka) [reply src/test/java/com/aws/greengrass/deployment/DeploymentTaskTest.java:95]: removing!

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java (hunk 8 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DeploymentService.java:366]: I think something got mixed up here! will look at it again.

File: src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java (hunk 7 above)
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/DeploymentDirectoryManager.java:88]: done

**Review by MikeDombo — DISMISSED** [review]
LGTM, let's see the CI pass and have another reviewer.

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java [hunk 11]
```diff
@@ -201,6 +204,25 @@ private static String parseGroupNameFromConfigurationArn(Configuration config) {
...
+
+        String groupName;
+        try {
+            // IoTJob Arn only gives thing group name without 'thinggroup/' prefix
+            groupName = Arn.fromString(arn).resource().resource();
+            if (groupName != null) {
+                groupName = new StringBuilder(THING_GROUP_RESOURCE_NAME_PREFIX).append(groupName).toString();
+            }
+        } catch (IllegalArgumentException e) {
+            // so that it can proceed, rather than fail, when the format of configurationArn is wrong.
```
[the remark this answers is not in the document]
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:220]: This is similar to configuration arn behavior. I was expecting this to be already handled, as the same thing can happen with a bad configuration arn.

Could you shed some light on why this would be any different?

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java [hunk 12]
```diff
@@ -154,7 +154,10 @@ public static DeploymentDocument convertFromDeploymentConfiguration(Configuratio
                         .deploymentId(config.getDeploymentId())
                         .requiredCapabilities(config.getRequiredCapabilities())
```
[the remark this answers is not in the document]
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:155]: Could you help me with the definition of required capabilities? 

I am unclear on what should be added as a required capability.

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:220]: It may have been a miss for existing configuration handling or maybe it didn't matter earlier since there was only one arn and as long as that was same across deployments, dependency resolution would work fine. Now there's an additional arn which has to be same as the group arn it's mimicking, if one is malformed/wrong then there will be a conflict - so you don't get the `onBehalfOf` behavior you expect, and customers won't know why there's a version conflict.

[the remark this answers is not in the document]
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:220]: Adding explicit exception then:

`throw new InvalidRequestException("Malformed value in deployment configuration for 'onBehalfOf'.");`

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/SubGroupDeploymentIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+import org.mockito.Mock;
+import software.amazon.awssdk.core.exception.SdkClientException;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (devnaruka) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/SubGroupDeploymentIntegrationTest.java:44]: removed

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:155]: Added SUBGROUP_DEPLOYMENTS required capability.

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java [hunk 14]
```diff
@@ -201,6 +205,36 @@ private static String parseGroupNameFromConfigurationArn(Configuration config) {
...
+                                              Configuration config) throws InvalidRequestException {
+        if (Utils.isNotEmpty(config.getOnBehalfOf())) {
+            try {
+                // IoT thingGroupArn only gives thing group name without 'thinggroup/' prefix
+                String groupName = Arn.fromString(config.getOnBehalfOf()).resource().resource();
+                if (groupName != null) {
+                    groupName = new StringBuilder(THING_GROUP_RESOURCE_NAME_PREFIX).append(groupName).toString();
+                }
+                deployDocBuilder.onBehalfOf(groupName);
+            } catch (IllegalArgumentException e) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:219]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 232.

File: src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java [hunk 15]
```diff
@@ -201,6 +205,36 @@ private static String parseGroupNameFromConfigurationArn(Configuration config) {
         return groupName;
     }
 
+    @SuppressWarnings("PMD.PreserveStackTrace")
+    private static void convertThingGroupArns(DeploymentDocument.DeploymentDocumentBuilder deployDocBuilder,
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:209]: I'm on mobile right now so it is a bit hard to see. What are we doing when both of the new fields are not present, which would be the case for all deployments today?

Comment (devnaruka) [reply src/main/java/com/aws/greengrass/deployment/converter/DeploymentDocumentConverter.java:209]: Nothing! If they are empty, we don't do arn parsing and the deployment would work as is.

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/deployment/FleetConfigWithSimpleAppv1AndGreenSignal.json [hunk 16]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (shaguptashaikh) [inline src/integrationtests/resources/com/aws/greengrass/integrationtests/deployment/FleetConfigWithSimpleAppv1AndGreenSignal.json:1]: You don't need the new required capability in the deployment doc for the tests to pass?

Comment (devnaruka) [reply src/integrationtests/resources/com/aws/greengrass/integrationtests/deployment/FleetConfigWithSimpleAppv1AndGreenSignal.json:1]: I am adding that in the test code. Check line `307` of `SubGroupDeploymentIntegrationTest`

**Review by shaguptashaikh — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]
